### PR TITLE
RI-463: Enable nested virtualization

### PR DESCRIPTION
**Pull request by coreywright**
Nested virtualization is required for KVM-based Nova instances to
successfully launch on VMs, accommodating RH OSP 13's default `kvm`
virtualization type and multi-node-AIO VM-based compute nodes.

### CODE REVIEW

**Review by antonym — APPROVED** [review]

**Comment by antonym** [comment]
LGTM, good to merge?

**Comment by coreywright** [comment]
nope, not ready to merge; it's still a "work in progress".  i'm in the process of developing tempest testing, would like to verify that tempest testing fails in the gate due to lack of nested virtualization, and then add this at that time to verify that it resolves the tempest test failures.

**Comment by antonym** [comment]
This good to go now too?

**Comment by coreywright** [comment]
@antonym,

tl;dr i'm cool with closing this PR and re-opening it later if it proves necessary for some particular use-case.

as i detailed in the associated JIRA issue, this shouldn't be necessary and should be taken care of by the underlying system packages, though i don't know why that didn't happen for me the first time which led to all this.

the only thing here that might be beneficial is adding `<feature policy='require' name='vmx'/>` to the VM template to require that nested virtualization is available (otherwise libvirt should produce an error on trying to create the VMs), instead of waiting until tempest testing and getting round-about error messages.

**Comment by antonym** [comment]
K, if we don't need it then we can close it down.  If we run into any issues, we can always reopen.

**Comment by antonym** [comment]
Ran into this today testing 12-13 upgrade, don't think this would actually hurt having in place.

**Comment by coreywright** [comment]
@antonym,

this is only problematic if:
1. ran on non-Intel hardware (eg AMD) as the `vmx` CPU flag is made a requirement in the libvirt VM template (and AMD has their own CPU flag for their hardware-assisted CPU virtualization)
2. QEMU is desired for CPU virtualization (as compared to KVM) because of lack of hardware support (regardless of Intel or AMD), which RH OSP accommodates, but doesn't support, though requires further configuration during deployment, IIRC (and there's a bug associated with QEMU usage in RH OSP 12 & 13)

**Comment by antonym** [comment]
recheck_all

**Comment by coreywright** [comment]
@antonym,

i've created branch [gating-periodic-daily-weekly](https://github.com/rcbops/osp-mnaio/tree/gating-periodic-daily-weekly) and while testing it (see [smoke-tests](https://rpc.jenkins.cit.rackspace.net/job/PM_osp-mnaio-master-xenial_mnaio-3ctlr_2comp_3ceph-osp_13_deploy-smoke-tests/) and [full-tests](https://rpc.jenkins.cit.rackspace.net/job/PM_osp-mnaio-master-xenial_mnaio-3ctlr_2comp_3ceph-osp_13_deploy-full-tests)) i encountered hosts failing due to `libvirtError: invalid argument: could not find capabilities for domaintype=kvm`, which i previously traced down to lack of nested virtualization.

i wish the problem was definitively repeatable to verify this resolves the problem (without modifying the playbooks and such to insert debug tasks to verify that *after* installing kvm-qemu the kvm-intel module is not inserted into the kernel, but after running this it is).

EDIT: ran the smoke-tests/daily job again and it failed the same (ie build 10), so i'm going to cherry-pick this PR's commits on to the gating-periodic-daily-weekly branch, run it a few times (starting with build 11, which i just kicked off), and see if it fails again.

**Comment by antonym** [comment]
I did change the OS in gating from Xenial to Bionic last week so maybe that’s the variable.

**Comment by coreywright** [comment]
> I did change the OS in gating from Xenial to Bionic last week so maybe that’s the variable.

but i've been using bionic personally this whole time (because that's what you originally prescribed), and i had fail once (which kicked off my whole research into it) and never again (though i'm only building once a week or less, not multiple times in a day like i'm currently doing in jenkins).

i did add the 2 commits from this PR and the job then passed (see [osp_13_deploy-smoke-tests build #12](https://rpc.jenkins.cit.rackspace.net/job/PM_osp-mnaio-master-xenial_mnaio-3ctlr_2comp_3ceph-osp_13_deploy-smoke-tests/12)) but would like to do repeat it (without running into the maintenance window) as it reported that nested virtualization was already enabled, but that definitely wasn't the case with the prior builds that failed.

```
Intel CPU found. Checking for nested virtualization capabilities.
Nested virtualization capability found. Checking if it is enabled.
Nested virtualization already enabled.
```

**Comment by coreywright** [comment]
see [RI-503](https://rpc-openstack.atlassian.net/browse/RI-503) for my results of using this PR's commits with osp-mnaio's Jenkins periodic job testing.

tl;dr where jobs failed due to lack of nested virtualization before cherry-picking this PR's commit, then one build done after the cherry-picking records that nested virtualization was already enabled (and didn't need the cherry-picks).

i'm going to gather more data running Jenkins builds.

**Comment by antonym** [comment]
We probably are fine to merge this in and use it then.  If we ever need to run this on AMD then we can adjust from there.

**Comment by antonym** [comment]
Ran into this on one of the gating tests where smoke failed, probably related to not being able to nest.  Merging this in and if it causes issues, we can revisit.
